sycl/doc/extensions/proposed/sycl_ext_oneapi_raw_kernel.asciidoc

```diff
@@ -0,0 +1,563 @@
+specification.*
+
+
+== Overview
+
+This extension introduces a new way to define a kernel as a simple C++
+function, where the kernel arguments are parameters to the function.
+This is different from standard SYCL kernels, where the kernel arguments are
+either captures of a lambda expression or member variables of a callable
+object.
+
+The primary motivation for this extension is the
+link:../proposed/sycl_ext_oneapi_kernel_compiler.asciidoc[
+sycl_ext_oneapi_kernel_compiler], which allows online compilation of a kernel
+from source code.
+These kernels must have a clean separation between host and device code, and
+they must have a defined order to their arguments because the application sets
+the argument values by their "index" via `handler::set_arg(index, value)`.
+Because variables captured by lambda expressions have no defined order, we need
+some other way to define kernel arguments for use with the
+sycl_ext_oneapi_kernel_compiler extension, and the sycl_ext_oneapi_raw_kernel
+extension provides that mechanism.
+The "raw kernel" feature is a separate extension, though, because it can also
+be used independently from sycl_ext_oneapi_kernel_compiler.
+For example, some users may find it more familiar to define kernels as plain
+functions or they might have other reasons to prefer a clean separation between
+host and device code.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.
+An implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_RAW_KERNEL`
+to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Defining a raw kernel
+
+A raw kernel is a normal C++ function definition, where the function definition
+is decorated with one of the following compile-time properties: `range_kernel`,
+`nd_range_kernel`, or `single_task_kernel`.
+Each of these properties takes a template parameter `Name`, which serves as the
+name of the kernel, similar to the `KernelName` template parameter to
+`parallel_for` for a standard SYCL kernel.
+When a raw kernel has a name, the type used as the name must not be used to
+name any other kernel in the application (either a raw kernel or a "named"
+standard kernel).
+The type used to name a raw kernel has the same limitations as the type used to
+name a standard SYCL kernel as specified in section 5.2 "Naming of kernels" of
+the core SYCL specification.
+A raw kernel may be defined as a function template, however, all instantiations
```

Comment:
Do we really need support `function template`? Can we keep it a "simple C++ function" as described in the overview and let user to provide function definitions explicitly?

From POV, supporting just C functions covers the need to provide well defined kernel ABI.
Function overloading via name mangling is an extra feature, which is not required to solve the problem this extension aims at. Certain features like function templates complicate the implementation of SYCL API and supporting tools (e.g. debuggers, profiles, etc.).

Reply:
I floated this idea in one of my presentations, and I got feedback that we should support even complex cases like templates and overloads.  Note that NVRTC supports kernels that are templates and overloads, and this "free function kernel" feature is part of our response to that feature -- a portable API that dynamically compiles kernels from a source string.

Reply:
https://docs.nvidia.com/cuda/nvrtc/index.html#accessing-lowered-names - NVRTC solution for supporting C++ functions. Their solution more straightforward: device compiler is expected to use standard C++ mangling to support function templates and overloading. User should provide exact name generated by the device compiler and NVIDIA provides a couple of helper functions to lower a readable function name to the kernel name emitted by the device compiler.

Do you think that our proposal (i.e. using type-based kernel identifier) for free function kernel has any advantages vs using string-based identifier?
It certainly does for a single source - the code is less error-prone as both compilers see a single use of the identifier. In separate compilation case, using type-based kernel identifier doesn't provide this advantage, but looks alien.

Reply:
I removed the support for identifying a kernel by its type-name and replaced it with a way to identify the kernel by its address in a7bef5ed34e9e71199da439fa5d27015782c191d.  I think this resolves this conversation?

I'll mark this resolved next week unless you add another comment.